### PR TITLE
Albrja/mic 5267/build from vbu

### PR DESCRIPTION
**Pull request by albrja**
### Add reusable workflow
- *Category*: CI
- *JIRA issue*: https://jira.ihme.washington.edu/browse/MIC-5267

### CODE REVIEW

File: .github/workflows/build.yml [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        required: true
+        type: string
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ["3.9", "3.10", "3.11"]
+    defaults:
```
Comment (stevebachmeier) [inline .github/workflows/build.yml:20]: What is this bit doing?

File: .github/workflows/build.yml [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+      - name: Update pip
+        run: |
+          python -m pip install --upgrade pip
+      - name: Install dependencies
+        run: |
+          pip install .[dev]
+          git checkout ${branch_name}
+      - name: Check for dependency branches and install if needed
+        run: |
+          git clone https://github.com/ihmeuw/vivarium_build_utils.git ../vivarium_build_utils
```
Comment (stevebachmeier) [inline .github/workflows/build.yml:47]: I don't think you need the `/vivarium_build_utils` after the `..` right? Woudn't that create two `vivarium_build_utils` folders?

File: .github/workflows/build.yml [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+          else
+            echo "branch_name=${GITHUB_REF_NAME}" >> $GITHUB_ENV
+          fi
+      - name: Update pip
+        run: |
+          python -m pip install --upgrade pip
+      - name: Install dependencies
+        run: |
+          pip install .[dev]
+          git checkout ${branch_name}
```
Comment (stevebachmeier) [inline .github/workflows/build.yml:44]: where does `${branch_name}` come from?

File: .github/workflows/build.yml [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+      - name: Check for dependency branches and install if needed
+        run: |
+          git clone https://github.com/ihmeuw/vivarium_build_utils.git ../vivarium_build_utils
+          echo "----------------------------------------"
+          echo "Contents of install_dependency_branch.sh"
+          echo "----------------------------------------"
+          cat ../vivarium_build_utils/install_dependency_branch.sh
+          echo "" 
+          echo "----------------------------------------"
+          for package in $(echo ${{ inputs.dependencies }} | sed "s/,/ /g"); do
```
Comment (stevebachmeier) [inline .github/workflows/build.yml:54]: nice

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+            sh ../vivarium_build_utils/install_dependency_branch.sh ${package} ${branch_name} github
+          done
+      - name: print environment values
+        run: |
+          cat $GITHUB_ENV
+      - name: Lint
+        run: |
+          pip install black==22.3.0 isort
```
Comment (stevebachmeier) [inline .github/workflows/build.yml:62]: You dropped mypy. We'll need to figure this one out while we transition. How are you figuring we'll handle it until we implement mypy in all of our frameworks?

**Review by stevebachmeier — CHANGES_REQUESTED** [review]
Super exciting! I do think we need to figure out how to customize repo stages, though, with the big obvious example being we can't just lose mypy for vivarium.

File: .github/workflows/build.yml (hunk 3 above)
Comment (albrja) [reply .github/workflows/build.yml:48]: It just knows from being called in vivarium.

File: .github/workflows/build.yml (hunk 2 above)
Comment (albrja) [reply .github/workflows/build.yml:51]: Yes I want to do that so we don't have to cd although I think this change was unnecessary. It is the same as cd .. git clone but less lines of code.

File: .github/workflows/build.yml (hunk 1 above)
Comment (albrja) [reply .github/workflows/build.yml:24]: Just setting the shell.

File: .github/workflows/build.yml [hunk 6]
```diff
@@ -0,0 +1,97 @@
+# -----------------------------------------------------------------------------
+#   - invoked on push, pull_request, manual trigger, or schedule
+#   - test under at least 3 versions of python
+# -----------------------------------------------------------------------------
+name: build
+on:
+  workflow_call:
+    inputs:
+      dependencies:
+        description: "String of comma-separated dependencies to check for upstream branches"
```
[the remark this answers is not in the document]
Comment (stevebachmeier) [reply .github/workflows/build.yml:10]: Jim confirmed no we can't pass lists in

**Review by stevebachmeier — APPROVED** [review]

File: .github/workflows/build.yml (hunk 6 above)
[the remark this answers is not in the document]
Comment (albrja) [reply .github/workflows/build.yml:10]: Github only allows bools, int, or strings. There should be no spaces. But we will only have to configure those once most likely in the other config files then we won't have to worry about it anymore.

**Review by hussain-jafari — APPROVED** [review]

**Pull request merged**